// @astryxdesign/core/dist/Calendar/dayCellUtils.js
// Copyright (c) Meta Platforms, Inc. and affiliates.

import { plainDateIsEqual, plainDateIsInRange } from "../utils/plainDate.js";
/** Derives all visual/interaction states for a single day cell. */
export function computeDayCellState(input) {
  const {
    date,
    dayIndex,
    mode,
    selectedDate,
    rangeStart,
    rangeEnd,
    previewStart,
    previewEnd,
    today,
    isDisabled,
    isOutside
  } = input;
  return {
    effectivelyDisabled: isDisabled || isOutside,
    isToday: plainDateIsEqual(date, today),
    isSelected: !!(mode === 'single' && selectedDate && plainDateIsEqual(date, selectedDate)),
    isInRange: !!(mode === 'range' && rangeStart && rangeEnd && plainDateIsInRange(date, [rangeStart, rangeEnd])),
    isRangeStart: !!(mode === 'range' && rangeStart && plainDateIsEqual(date, rangeStart)),
    isRangeEnd: !!(mode === 'range' && rangeEnd && plainDateIsEqual(date, rangeEnd)),
    isInPreview: !!(previewStart && previewEnd && plainDateIsInRange(date, [previewStart, previewEnd])),
    isPreviewStart: !!(previewStart && plainDateIsEqual(date, previewStart)),
    isPreviewEnd: !!(previewEnd && plainDateIsEqual(date, previewEnd)),
    isFirstColumn: dayIndex === 0,
    isLastColumn: dayIndex === 6
  };
}

/** Edge rounding for range background (rounds at grid edges and range endpoints). */
export function computeRangeRounding(state) {
  return {
    roundLeft: state.isRangeStart || state.isFirstColumn,
    roundRight: state.isRangeEnd || state.isLastColumn
  };
}

/** Edge rounding for preview background. */
export function computePreviewRounding(state) {
  return {
    roundLeft: state.isPreviewStart || state.isFirstColumn,
    roundRight: state.isPreviewEnd || state.isLastColumn
  };
}

/** Whether a day is a selection endpoint (selected, range start, or range end). */
export function isEndpoint(state) {
  return state.isSelected || state.isRangeStart || state.isRangeEnd;
}
